File: kh3_src/tabs/general.py
from dash import Dash, html, dcc, callback, Input, Output, State, ALL
import kh3_src.kh3_utils as utils


def get_playtime(playtime):
    hours = playtime // 3600
    minutes = (playtime % 3600) // 60
    seconds = (playtime % 3600) % 60
    return hours, minutes, seconds

def calculate_playtime(hours, minutes, seconds):
    return hours * 3600 + minutes * 60 + seconds

def create_general():
    kh3 = utils.kh3
    hours, minutes, seconds = get_playtime(kh3.playtime.value)
    playtime = html.Div([
        dcc.Input(
            id="Hours",
            type="number",
            value=hours,
            min=0,
            max=399,
            step=1,
            style={"width": 50},
        ),
        html.Label(" : "),
        dcc.Input(
            id="Minutes",
            type="number",
            value=minutes,
            min=0,
            max=59,
            step=1,
            style={"width": 50},
        ),
        html.Label(" : "),
        dcc.Input(
            id="Seconds",
            type="number",
            value=seconds,
            min=0,
            max=59,
            step=1,
            style={"width": 50},
        ),
    ])
    desire = dcc.Dropdown(
        options=[
            {"label": "Vitality", "value": 0},
            {"label": "Wisdom", "value": 1},
            {"label": "Balance", "value": 2},
        ],
        value=kh3.desire.value,
        id="Desire",
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    power = dcc.Dropdown(
        options=[
            {"label": "Warrior", "value": 0},
            {"label": "Mystic", "value": 1},
            {"label": "Guardian", "value": 2},
        ],
        value=kh3.power.value,
        id="Power",
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    leader = dcc.Dropdown(
        options=[
            {"label": k, "value": v} for k, v in kh3.character_dict.items()
        ],
        value=kh3.party[0],
        id="Leader",
        disabled=False,
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    friend1 = dcc.Dropdown(
        options=[
            {"label": k if k != "Sora" else "Default", "value": v} for k, v in kh3.character_dict.items()
        ],
        value=kh3.party[1],
        id="Friend1",
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    friend2 = dcc.Dropdown(
        options=[
            {"label": k if k != "Sora" else "Default", "value": v} for k, v in kh3.character_dict.items()
        ],
        value=kh3.party[2],
        id="Friend2",
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    friend3 = dcc.Dropdown(
        options=[
            {"label": k if k != "Sora" else "Default", "value": v} for k, v in kh3.character_dict.items()
        ],
        value=kh3.party[3],
        id="Friend3",
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    friend4 = dcc.Dropdown(
        options=[
            {"label": k if k != "Sora" else "Default", "value": v} for k, v in kh3.character_dict.items()
        ],
        value=kh3.party[4],
        id="Friend4",
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    munny = dcc.Input(
        id="Munny",
        type="number",
        value=kh3.munny.value,
        min=0,
        max=0xFFFFFFFF,
        step=1,
        style={"width": 100},
    )
    exp = dcc.Input(
        id="EXP",
        type="number",
        value=kh3.exp.value,
        min=0,
        max=0xFFFFFFFF,
        step=1,
        style={"width": 100},
    )
    difficulty = dcc.Dropdown(
        options=[
            {"label": "Beginner", "value": 0},
            {"label": "Standard", "value": 1},
            {"label": "Proud", "value": 2},
            {"label": "Critical", "value": 3},
        ],
        value=kh3.difficulty.value,
        id="Difficulty",
        searchable=False,
        clearable=False,
        style={"width": 200},
    )
    map_path = html.Div([
        dcc.Input(
            id="Map Path",
            type="text",
            value=kh3.map_path[:kh3.map_path.find(0)].decode("utf-8"),
            style={"width": "95%"},
        ),
    ],
        style={"margin-top": 10},
    )
    map_spawn = html.Div([
        dcc.Input(
            id="Map Spawn",
            type="text",
            value=kh3.map_spawn[:kh3.map_spawn.find(0)].decode("utf-8"),
            style={"width": "95%"},
        ),
    ],
        style={"margin-top": 10},
    )
    player_script = html.Div([
        dcc.Input(
            id="Player Script",
            type="text",
            value=kh3.player_script[:kh3.player_script.find(0)].decode("utf-8"),
            style={"width": "95%"},
        ),
    ],
        style={"margin-top": 10},
    )
    player_pawn = html.Div([
        dcc.Input(
            id="Player Pawn",
            type="text",
            value=kh3.player_pawn[:kh3.player_pawn.find(0)].decode("utf-8"),
            style={"width": "95%"},
        ),
    ],
        style={"margin-top": 10},
    )
    return html.Div([
        html.Div([dcc.Markdown("Playtime:"), playtime]),
        html.Div([dcc.Markdown("Desire:"), desire]),
        html.Div([dcc.Markdown("Power:"), power]),
        html.Div([dcc.Markdown("Party:"), leader, friend1, friend2, friend3, friend4]),
        html.Div([dcc.Markdown("Munny:"), munny]),
        html.Div([dcc.Markdown("EXP:"), exp]),
        html.Div([dcc.Markdown("Difficulty:"), difficulty]),
        html.H3("Advanced Options"),
        html.H4("Room Mod"),
        html.Div([dcc.Markdown("Map Path:"), map_path]),
        html.Div([dcc.Markdown("Map Spawn:"), map_spawn]),
        html.H4("Player Mod"),
        html.Div([dcc.Markdown("Player Script:"), player_script]),
        html.Div([dcc.Markdown("Player Pawn:"), player_pawn]),
    ])

@callback(
    Input("Hours", "value"),
    Input("Minutes", "value"),
    Input("Seconds", "value"),
)
def playtime_callback(
    hours,
    minutes,
    seconds
):
    kh3 = utils.kh3
    try:
        kh3.playtime.value = calculate_playtime(hours, minutes, seconds)
    except:
        pass

@callback(
    Input("Desire", "value"),
    Input("Power", "value"),
    Input("Leader", "value"),
    Input("Friend1", "value"),
    Input("Friend2", "value"),
    Input("Friend3", "value"),
    Input("Friend4", "value"),
    Input("Munny", "value"),
    Input("EXP", "value"),
    Input("Difficulty", "value"),
    Input("Map Path", "value"),
    Input("Map Spawn", "value"),
    Input("Player Script", "value"),
    Input("Player Pawn", "value"),
)
def general_callback(
    desire,
    power,
    leader,
    friend1,
    friend2,
    friend3,
    friend4,
    munny,
    exp,
    difficulty,
    map_path,
    map_spawn,
    player_script,
    player_pawn,
):
    kh3 = utils.kh3
    try:
        kh3.desire.value = desire
        kh3.power.value = power
        kh3.party[0] = leader
        kh3.party[1] = friend1
        kh3.party[2] = friend2
        kh3.party[3] = friend3
        kh3.party[4] = friend4
        kh3.munny.value = munny
        kh3.exp.value = exp
        kh3.difficulty.value = difficulty
    except:
        pass
    if len(map_path) <= 0x100:
        kh3.map_path = bytearray(map_path, "utf-8") + bytearray(0x100 - len(map_path))
    if len(map_spawn) <= 0x40:
        kh3.map_spawn = bytearray(map_spawn, "utf-8") + bytearray(0x40 - len(map_spawn))
    if len(player_script) <= 0x100:
        kh3.player_script = bytearray(player_script, "utf-8") + bytearray(0x100 - len(player_script))
    if len(player_pawn) <= 0x100:
        kh3.player_pawn = bytearray(player_pawn, "utf-8") + bytearray(0x100 - len(player_pawn))

File: kh3_src/tabs/general.js
export function create_general() {
    const kh3div = document.getElementById("kh3div");
    const playtime = create_playtime();
    const power = `
    <div>
        <h3>Power:</h3>
        <select id="power">
            <option value=0>Warrior</option>
            <option value=1>Mystic</option>
            <option value=2>Guardian</option>
        </select>
    </div>`;
    const desire = `
    <div>
        <h3>Desire:</h3>
        <select id="desire">
            <option value=0>Vitality</option>
            <option value=1>Wisdom</option>
            <option value=2>Balance</option>
        </select>
    </div>`;
    const munny = `
    <div>
        <h3>Munny:</h3>
        <input
            type="number"
            id="munny"
            min=0
            max=${0xFFFFFFFF}
            step=1
            value=${window.kh3.munny.value}
        >
    </div>`;
    const exp = `
    <div>
        <h3>EXP:</h3>
        <input
            type="number"
            id="exp"
            min=0
            max=${0xFFFFFFFF}
            step=1
            value=${window.kh3.exp.value}
        >
    </div>`;
    const difficulty = `
    <h3>Difficulty</h3>
    <select id="difficulty">
        <option value=0>Beginner</option>
        <option value=1>Standard</option>
        <option value=2>Proud</option>
        ${window.kh3.version > 0 ? "<option value=3>Critical</option>" : ""}
    </select>`;
    const decoder = new TextDecoder("utf-8");
    const path = decoder.decode(new Uint8Array(window.kh3.map_path));
    const map_path = `
    <div>
        <h4>Map Path:</h4>
        <input
            type="text"
            id="map_path"
            value="${path.slice(0, path.indexOf('\0'))}"
            style="width: 500px"
        >
    </div>`;
    const spawn = decoder.decode(new Uint8Array(window.kh3.map_spawn));
    const map_spawn = `
    <div>
        <h4>Map Spawn:</h4>
        <input
            type="text"
            id="map_spawn"
            value="${spawn.slice(0, spawn.indexOf('\0'))}"
            style="width: 500px"
        >
    </div>`;
    const script = decoder.decode(new Uint8Array(window.kh3.player_script));
    const player_script = `
    <div>
        <h4>Player Script:</h4>
        <input
            type="text"
            id="player_script"
            value="${script.slice(0, script.indexOf('\0'))}"
            style="width: 500px"
        >
    </div>`;
    const pawn = decoder.decode(new Uint8Array(window.kh3.player_pawn));
    const player_pawn = `
    <div>
        <h4>Player Pawn:</h4>
        <input
            type="text"
            id="player_pawn"
            value="${pawn.slice(0, pawn.indexOf('\0'))}"
            style="width: 500px"
        >
    </div>`;
    kh3div.innerHTML = `
    <div>
        ${playtime}
        ${power}
        ${desire}
        ${munny}
        ${exp}
        ${difficulty}
        <h3>Advanced Options</h3>
        <h3>Room Mod</h3>
        ${map_path}
        ${map_spawn}
        <h3>Player Mod</h3>
        ${player_script}
        ${player_pawn}
    </div>`;
    general_callbacks();
}

function create_playtime() {
    const playtime = window.kh3.get_playtime(window.kh3.playtime.value);
    return `
    <h3>Playtime:</h3>
    <div id="playtime" style="display: flex; gap: 20px">
        <input
            type="number"
            id="hours"
            min=0
            max=400
            step=1
            value=${playtime[0]}
        >
        <input
            type="number"
            id="minutes"
            min=0
            max=59
            step=1
            value=${playtime[1]}
        >
        <input
            type="number"
            id="seconds"
            min=0
            max=59
            step=1
            value=${playtime[2]}
        >
    </div>`;
}

function playtime_callbacks() {
    const hours = document.getElementById("hours");
    const minutes = document.getElementById("minutes");
    const seconds = document.getElementById("seconds");
    function update_playtime() {
        if (
            hours.validity.valid &&
            minutes.validity.valid &&
            seconds.validity.valid
        ) {
            window.kh3.playtime.value = window.kh3.set_playtime(
                Number(hours.value),
                Number(minutes.value),
                Number(seconds.value),
                Number(0),
            );
        }
        const [h, m, s, f, cs] = window.kh3.get_playtime(window.kh3.playtime.value);
        hours.value = h;
        minutes.value = m;
        seconds.value = s;
    }
    hours.addEventListener("change", update_playtime);
    minutes.addEventListener("change", update_playtime);
    seconds.addEventListener("change", update_playtime);
}

function general_callbacks() {
    playtime_callbacks();
    const power = document.getElementById("power");
    power.value = window.kh3.power.value;
    power.addEventListener("change", () => {
        window.kh3.power.value = power.value;
    });
    const desire = document.getElementById("desire");
    desire.value = window.kh3.desire.value;
    desire.addEventListener("change", () => {
        window.kh3.desire.value = desire.value;
    });
    const munny = document.getElementById("munny");
    munny.addEventListener("change", () => {
        if (munny.validity.valid)
            window.kh3.munny.value = munny.value;
        munny.value = window.kh3.munny.value;
    });
    const exp = document.getElementById("exp");
    exp.addEventListener("change", () => {
        if (exp.validity.valid)
            window.kh3.exp.value = exp.value;
        exp.value = window.kh3.exp.value;
    });
    const difficulty = document.getElementById("difficulty");
    difficulty.value = window.kh3.difficulty.value;
    difficulty.addEventListener("change", () => {
        window.kh3.difficulty.value = difficulty.value;
    });
    const decoder = new TextDecoder("utf-8");
    const encoder = new TextEncoder("utf-8");
    const map_path = document.getElementById("map_path");
    map_path.addEventListener("change", () => {
        window.kh3.map_path.fill(0);
        for (let i = 0; i < map_path.value.length; i++) {
            window.kh3.map_path[i] = encoder.encode(map_path.value)[i];
        }
        const path = decoder.decode(new Uint8Array(window.kh3.map_path));
        map_path.value = path.slice(0, path.indexOf('\0'));
    });
    const map_spawn = document.getElementById("map_spawn");
    map_spawn.addEventListener("change", () => {
        window.kh3.map_spawn.fill(0);
        for (let i = 0; i < map_spawn.value.length; i++) {
            window.kh3.map_spawn[i] = encoder.encode(map_spawn.value)[i];
        }
        const spawn = decoder.decode(new Uint8Array(window.kh3.map_spawn));
        map_spawn.value = spawn.slice(0, spawn.indexOf('\0'));
    });
    const player_script = document.getElementById("player_script");
    player_script.addEventListener("change", () => {
        window.kh3.player_script.fill(0);
        for (let i = 0; i < player_script.value.length; i++) {
            window.kh3.player_script[i] = encoder.encode(player_script.value)[i];
        }
        const script = decoder.decode(new Uint8Array(window.kh3.player_script));
        player_script.value = script.slice(0, script.indexOf('\0'));
    });
    const player_pawn = document.getElementById("player_pawn");
    player_pawn.addEventListener("change", () => {
        window.kh3.player_pawn.fill(0);
        for (let i = 0; i < player_pawn.value.length; i++) {
            window.kh3.player_pawn[i] = encoder.encode(player_pawn.value)[i];
        }
        const pawn = decoder.decode(new Uint8Array(window.kh3.player_pawn));
        player_pawn.value = pawn.slice(0, pawn.indexOf('\0'));
    });
}
